pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/FDAwareInstancePartitionSelector.java

```diff
@@ -0,0 +1,455 @@
+ * partition, with each fault-domain mapping 1:1 to a server pool
+ * Algorithm details and proof see
+ * https://docs.google.com/document/d/1KmJ1DsYXVdzrojj_JYBHRJ2gRMQ5y-o63YqPs7ei7nI/edit#heading=h.nrxxvujq7ael
+ */
+public class FDAwareInstancePartitionSelector extends InstancePartitionSelector {
+  private static final Logger LOGGER = LoggerFactory.getLogger(FDAwareInstancePartitionSelector.class);
+
+  public FDAwareInstancePartitionSelector(InstanceReplicaGroupPartitionConfig replicaGroupPartitionConfig,
+      String tableNameWithType, @Nullable InstancePartitions existingInstancePartitions) {
+    super(replicaGroupPartitionConfig, tableNameWithType, existingInstancePartitions);
+  }
+
+  /**
+   * @return pair of (numFaultDomains, numTotalInstances)
+   */
+  private Pair<Integer, Integer> processFaultDomainPreconditions(
+      Map<Integer, List<InstanceConfig>> faultDomainToInstanceConfigsMap) {
+    int numFDs = faultDomainToInstanceConfigsMap.size();
+    Preconditions.checkState(numFDs != 0, "No pool (fault-domain) qualified for selection");
+
+    // Collect the number of instances in each FD into a list and sort
+    List<Integer> numInstancesPerFD =
+        faultDomainToInstanceConfigsMap.values().stream().map(List::size).sorted().collect(Collectors.toList());
+    // Should have non-zero total instances
+    Optional<Integer> totalInstancesOptional = numInstancesPerFD.stream().reduce(Integer::sum);
+    Preconditions.checkState(totalInstancesOptional.orElse(0) > 0, "The number of total instances is zero");
+    int numTotalInstances = totalInstancesOptional.get();
+    // Assume best-effort round-robin assignment of instances to FDs, the assignment should be balanced
+    // i.e., the difference between max num instances and min num instances should be at most 1
+    Preconditions.checkState(numInstancesPerFD.get(numInstancesPerFD.size() - 1) - numInstancesPerFD.get(0) <= 1,
+        "The instances are not balanced for each pool (fault-domain)");
+    return new ImmutablePair<>(numFDs, numTotalInstances);
+  }
+
+  /**
+   * @return (numReplicaGroups, numInstancesPerReplicaGroup)
+   */
+  private Pair<Integer, Integer> processReplicaGroupAssignmentPreconditions(int numFaultDomains, int numTotalInstances,
+      InstanceReplicaGroupPartitionConfig replicaGroupPartitionConfig) {
+    int numReplicaGroups = replicaGroupPartitionConfig.getNumReplicaGroups();
+    Preconditions.checkState(numReplicaGroups > 0, "Number of replica-groups must be positive");
+    int numInstancesPerReplicaGroup = replicaGroupPartitionConfig.getNumInstancesPerReplicaGroup();
+    if (numInstancesPerReplicaGroup > 0) {
+      int numInstancesToSelect = numInstancesPerReplicaGroup * numReplicaGroups;
+
+      Preconditions.checkState(numInstancesToSelect <= numTotalInstances,
+          "Not enough qualified instances (%s in all FDs, asked for %s)", numTotalInstances, numInstancesToSelect);
```

Comment:
How about rephrasing the sentence to `ask for: (numInstancesPerReplicaGroup: %d) *  (numReplicaGroups: %d) = %d` to give better explanation?